feat(connector v3 docs): adds documentation of new v3 features (#23)
provable-things/docs: source/includes/_ethereum.md

uint256 newGasPrice = <new-gas-price>;

oraclize_requestCallbackRebroadcast(
  <query-id>,
  newGasLimit,
  newGasPrice
  oraclize_getRebroadcastCost(newGasLimit, newGasPrice)
);
```

This way, the effects of the changing network conditions can be mitigated, ensuring callbacks are always received in a timely manner.

Another useful feature of rebroadcasting is in giving a user a finer-control over the gas limit of their callback. Should a query be sent with an amount of gas too low for the `__callback()` to fully execute, the query would normally `revert()` and be wasted. Now a user has the ability to request a rebroadcast on that query and thus the opportunity to make changes to the gas limit in order to remedy the situation.

In order to enable rebroadcasts, a contract writer must explicitly define the following in the storage of their contract: `bool public constant allowQueryRebroadcasts = true;`. Any attempts to rebroadcast a query destined for a contract which does not have the preceding line present will be unfulfilled.

Just like a normal query, any over-payment is refunded, but always to the _final calling contract_. Which latter gives rise to two important caveats:

> - The first is that the refund will _always_ go to the final calling contract, and so if the call for a rebroadcast is proxied via another contract, the refund will always be made to the that implements the API, and not necessarily to the contract which initiated and paid for the call.

Reviewer: to the contract that implements the API (contract is missing I take it?)

Author: Oopsie!

> - The second caveat is that the final calling contract will also require a payable `fallback()` function in order to receive refunds in such cases. Absent this function, the refund attempt will cause the query to `revert();`.

Reviewer: ```suggestion
- The second caveat is that the final calling contract will also require a payable `fallback()` function in order to receive refunds in such cases. If the payable fallback is omitted, any refund attempt will cause the query to `revert();`.
```

I think this is a bit clearer, it took me a few tries to read `Absent this function, ...` I also know you love `if` statements ;)

I was also wondering whether to add another point exclaiming the point that it is fine, as long as the payment is exact, but I should assume a decent level of intelligence and reading comprehension by th reader, as it's mentioned previously.

Reviewer: speak of the devil at L291, perfect

Author: But `absent` as a preposition reads so nicely :sob:

```

Payment for Provable queries are debited directly from the contract that calls the `oraclize_query(...)'` function at the moment that function call is made. So - depending on your contract's logic - you may want to know ahead of time the price of your next query before making it. Provable provides a helper method to do just that: `oraclize_getPrice(string _datasource);`.

> 

Reviewer: regarding the piggies article, it may be worth to maybe point out 'atomic-like' support, where the contract balance is empty, and needs to be funded by users each time (effectively how etheroll workds too)

Author: Yeah good point - super timely in fact. Will update to reflect that.

### Encrypted Queries
Certain contexts, such as smart contracts on public blockchains, might require a level of privacy to protect data from public scrutiny. Developers can make encrypted Oraclize queries by encrypting a part (or all) of a query with the Oraclize public key.

> The encrypted queries feature may be of interested to developers who want to deploy their blockchain applications of public networks. For example, if an application leverages data from an authenticated API, it would be dangerous to disclose the API key to anyway who is monitoring the public chain.

Reviewer: ```suggestion
The encrypted queries feature may be of interest to developers who want to deploy their blockchain applications on public networks. For example, if an application leverages data from an authenticated API, it may be dangerous to disclose the API key to anybody who is monitoring the public chain.
```

Reviewer: pssht, and English is your first and only language, come on Grgorsz

Author: Shh!! I already feel inadequate in an office full of polyglots!!

Also, how on earth do you pronounce `Grgorsz`?

#### Multi-Party Interactions

In the case of multi-party interactions, such as voting schemes or lotteries, the commitment data can should include all participants addresses, to ensure that the transaction cannot be replayed by a miner on a fork or a reorged chain where a participant didn't put a stake.

### ERC20 Token Payments

```javascript

/**
 *
 * @dev Provable helper functions for managing ERC20 Payments
 *
 */

// Set payment method to token existing at <token-address>
oraclize_setCustomTokenPayment(address <token-address>);

// Approve an allowance of the token for the Provable service:
oraclize_approveTokenAllowance(
  address <token-address>,
  uint256 <token-amount>
);

// Or combine the above into a single call to this function:
oraclize_setAndApproveTokenAllowance(
 	address<token-address>,
	uint256<token-amount>
);

// Get the price of a query in tokens:
oraclize_getPriceERC20(uint256 <gas-limit>);

// Revert back to paying with ETH and revoke the approved token amount:
oraclize_unsetAndRevokeCustomPayment();

```

> Provable supports ERC20 token payments. In order to make queries using a token as payment rather than ETH, you must somewhere in your contract call: `oraclize_setCustomTokenPayment(address <token-address>);`, where `<token-address>` is any of those existing on the whitelist<!-- FIME Link here! --> of supported tokens.

Reviewer: FIXME needs fixing

maybe platform info is useful for this?

Author: Yeah, perfect candidate for `platform-info`. Given that a whitelist currently doesn't exist yet, I'll leave it as `coming-soon` or something in the docs.

</aside>

Just as when __[paying for queries using ETH](http://docs.oraclize.it/#ethereum-best-practices-pre-calculating-the-query-price)__, when paying using a token it is also possible to pre-calculate a query-price via: `oraclize_getPriceERC20(<datasource>);`. This returns the next query price expressed in units of the token that the your contract has set as its `customTokenPayment`.

Additional parameters may be passed to get prices for more specific query types. See the information to the right for a full list of `oraclize_getPriceERC20` function overloads. All overloaded versions of the `oraclize_getPrice` helper function have the `view` visibility specifier and so are free for your contract to call.

<aside class="notice">
Note that the first query to the Provable service is _free_, and so the first call to `oraclize_getPriceERC20` from any contract with a gas limit and/or gas price that is _less than or equal_ to the default values will result in the function _correctly_ returning `0`.
</aside>

> In the case of a contract specifying a token payment, but sending enough ETH to cover the cost of the query, ETH will be used as payment rather than the token.

Reviewer: ah, this is interesting and i do not recall it

Author: Yup: __[behaviour tested here](https://github.com/oraclize/connector-update-priv/blob/339db47b9c16aca36a59ce40c61948f7352443e8/test/connector/making-queries.js#L975)__.

Reviewer: kinda like having our facebook status saying: 

> Looking for: Whatever I can get

}
```

### Byte Datasources

As a more advanced feature, Provable are providing all datasource `strings` as type `bytes1` as well as `strings`. The strings remain more readable (`URL`, `Random` and so on…) in Provable queries, but as with any string-related manipulation in Solidity, using them costs more gas-wise.  As such, by offering the same set of datasources as a `bytes1` equivalent, advanced users may choose to declare their datasource in this manner instead and thus benefit from cheaper query gas-costs.

<aside class="notice">
All of Provable's API functions where a `string _datasource` parameter is used can have the parameter substituted for its `bytes1` equivalent without altering the behaviour of the function in any way. See the section on the right for the `string -> bytes1` datasource conversion dictionary.
</aside>

``` javascript

// Datasource strings and their `bytes1` equivalents:

|---------------------------|
|       `string` | `bytes1` |

> |---------------------------|

Reviewer: I'm fine with it as is, but it would make more sense to see them in desc/asc bytes1 order.

Author: Common-sense beats pyramid, so done.
